### ISSUE TITLE
Bringing ReasonReact reducers to React.js

### ISSUE THREAD

**Issue by mxstbr** [facebook/react#10596]
It'd be awesome if React.js had reducers, similar to what [ReasonReact](https://reasonml.github.io/reason-react/blog.html#reducers-are-here) recently introduced.

> We’re also decided to iterate on ReasonReact instead of core React, partially because we’re Reason users and partially because ReasonReact’s nature of being new & less used allow us to iterate exciting new APIs without legacy concerns 

Since I [couldn't find](https://github.com/facebook/react/issues?utf8=%E2%9C%93&q=is%3Aissue%20reducers) any previous talk about this I'm assuming this has been discussed internally, it'd be great to chat about it in the open and iterate on core React state management!

<details>
  <summary>The ReasonReact reducers example</summary>

```Reason
type actions =
  | BumpCount
  | ToggleDialog bool;

type state = {count: int, showDialog: bool};

let component = ReasonReact.reducerComponent "MyComponent";

let make ::className _children => {
  ...component,
  initialState: fun () => {count: 0, showDialog: false},
  reducer: fun action state =>
    switch action {
    | BumpCount => ReasonReact.Update {...state, count: state.count + 1}
    | ToggleDialog show => ReasonReact.Update {...state, showDialog: show}
    },
  render: fun {state, reduce} => {
    let message = "You've clicked " ^ string_of_int state.count ^ " time(s)";
    <div className>
      <button onClick=(reduce (fun _ => BumpCount))>
        (ReasonReact.stringToElement message)
      </button>
      <MyButton onToggle=(reduce (fun toggleState => ToggleDialog toggleState)) />
      (state.showDialog ? <Dialog /> : ReasonReact.nullElement)
    </div>
  }
};
```

</details>

**Comment by alexeyraspopov** [comment]
I just post it here:
https://github.com/reactjs/react-future/blob/master/09%20-%20Reduce%20State/01%20-%20Declarative%20Component%20Module.js

**Comment by acdlite** [comment]
Let's move discussion of new component APIs into the React Future repo.

> it'd be great to chat about it in the open and iterate on core React state management

For more context: we're holding off on any major additions or changes to the component API until [async rendering](https://github.com/facebook/react/issues/8830) is more stable. There are many new constraints and quirks created by async rendering. It would be counterproductive for us to introduce new APIs now, only to find out later that those APIs aren't resilient to async behaviors.

@trueadm is also experimenting with a React component compiler, which will likely also influence the design of the new API.

tl;dr: Change is coming, but not right now.

(And yes, we're in constant collaboration with the Reason team, including in the design of the new reducer API.)

**Comment by acdlite** [comment]
Related: https://github.com/facebook/react/issues/10580 and https://github.com/facebook/react/issues/10581
